torch_xla/experimental/distributed_checkpoint/manager.py

```diff
@@ -77,14 +96,35 @@ def __init__(self,
             Default: 1, which only allows a single async checkpoint to be
             pending at a time.
     """
-    raise NotImplementedError
+    assert dist.is_initialized(), "A process group is required."
+    assert save_period > 0, "save_period must be positive"
+    assert async_queue_size > 0, "async_queue_size must be positive"
+    assert max_to_keep >= 0, "max_to_keep must be non-negative"
+
+    self.base_path = path
+    self.save_period = save_period
+    self.max_to_keep = max_to_keep
+
+  def _get_path(self, step: int) -> str:
+    return os.path.join(self.base_path, str(step))
+
+  def _release_oldest_checkpoints(self):
+    if self.max_to_keep > 0:
+      tracked_steps = sorted(self.all_steps(), reverse=True)
```

Comment:
Hmm, in python, we can't just have a queue structure?

Reply:
The `all_steps` method returns the tracked steps from the fsspec backend, so I'm not directly tracking steps in the CheckpointManager. `tracked_steps` will hold all steps which exist on the persistent storage in decreasing order, so the earliest steps are at the end of the list to enable O(1) popping.

Reply:
It seems this will be a hot path? Basically, each time a ckpt is saved. You have O(N) to grep the all_steps and O(nlogn) to sort it. Then another O(N) to pop the items. It seems only the last O(N) is necessary.